### PR TITLE
Attempts to resolve DatabaseBusy errors

### PR DESCRIPTION
**Pull request by maackle**
- Put all transaction access behind a fair-locking `parking_lot::RwLock` to avoid starvation
- Add traces to show mutex locking and transaction lifetimes
- Retry commits on database write errors with exponential backoff
    - We can remove this if we never see a retry happen. This is just to gather more information in case it helps.

### CODE REVIEW

File: crates/holochain_sqlite/src/db.rs [hunk 1]
```diff
@@ -201,43 +207,109 @@ pub trait WriteManager<'e> {
...
-        Ok(result)
+        let result = f(txn);
+        tracing::trace!("Released read lock on TXN_MUTEX.");
+        result
+    }
+}
+
+impl<'e> WriteManager<'e> for PConn {
+    #[tracing::instrument(skip(self, f))]
+    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
```
Comment (neonphog) [inline crates/holochain_sqlite/src/db.rs:257]: I know we've been kinda deferring dealing with all the synchronous sql code vs tokio, but both the RwLocks and the backoff retry add additional layers that should be in spawn_blocking blocks and `with_reader` and `with_commit` made `async`.

File: crates/holochain_sqlite/src/db.rs [hunk 2]
```diff
@@ -201,43 +207,109 @@ pub trait WriteManager<'e> {
...
+    fn with_reader<E, R, F>(&'e mut self, f: F) -> Result<R, E>
     where
         E: From<DatabaseError>,
-        F: 'e + FnOnce(&mut Transaction) -> Result<R, E>,
+        F: 'e + FnOnce(Transaction) -> Result<R, E>,
     {
-        let mut txn = self
-            .transaction_with_behavior(TransactionBehavior::Exclusive)
+        tracing::trace!("Attempting to acquire read lock on TXN_MUTEX.");
+        let _g = TXN_MUTEX.read();
```
Comment (freesig) [inline crates/holochain_sqlite/src/db.rs:244]: There's no reason to lock for a reader. Readers don't affect writers

File: crates/holochain_sqlite/src/db.rs [hunk 3]
```diff
@@ -201,43 +207,109 @@ pub trait WriteManager<'e> {
...
+            max_elapsed_time: Some(Duration::from_secs(5)),
+            ..Default::default()
+        };
+
+        // NB: The `?` here converts all errors into `backoff::Error::Transient`.
+        //     If there are unrecoverable errors, we can explicitly map those
+        //     to `backoff::Error::Permanent`.
+        let attempt = || {
+            tracing::trace!("Attempting to acquire write lock on TXN_MUTEX.");
+            let _g = TXN_MUTEX.write();
```
Comment (freesig) [inline crates/holochain_sqlite/src/db.rs:279]: You can't use a static because that makes all writes to all databases serial. We want this per database.

**Review by freesig — CHANGES_REQUESTED** [review]

File: crates/holochain_sqlite/src/db.rs (hunk 3 above)
Comment (maackle) [reply crates/holochain_sqlite/src/db.rs:279]: Oh yeah of ourse

File: crates/holochain_sqlite/src/db.rs (hunk 1 above)
Comment (maackle) [reply crates/holochain_sqlite/src/db.rs:267]: Sounds like we decided to asyncify things in a followup PR

File: crates/holochain_sqlite/src/db.rs [hunk 4]
```diff
@@ -201,43 +219,113 @@ pub trait WriteManager<'e> {
...
+            );
+            let mx = get_txn_mutex(self.db_kind());
+            let _g = mx.lock();
+            tracing::trace!("Acquired lock on txn mutex for {:?}.", self.db_kind());
+
+            let mut txn = self
+                .transaction_with_behavior(TransactionBehavior::Exclusive)
+                .map_err(DatabaseError::from)
+                .map_err(E::from)?;
+            let result = (f.clone())(&mut txn)?;
```
Comment (freesig) [inline crates/holochain_sqlite/src/db.rs:300]: I'm not sure how big the data in the closure might be but seems this strategy forces a clone even if there's no contention. Might be a bit of a heavy lift if you're writing large items like a dna to the db?

**Review by freesig — APPROVED** [review]
Nice. I think this will work but not sure about the overhead on the retry. I think we should maybe add a todo to remove it eventually?
Another way to test if there are any `SqlBusy` errors that aren't from the busy_timeout running out would be to just up the busy timeout to some very large number.

File: crates/holochain_sqlite/src/db.rs (hunk 4 above)
Comment (maackle) [reply crates/holochain_sqlite/src/db.rs:300]: I definitely want to remove the retry (and clones) if the mutex fixes the problem. But I think it's better to add as many remedies as possible since we our testing time with Holo is precious. The retries are just a big ol' safety net for this next test.

I still maintain that it's possible that there are some edge cases in sqlite that we aren't aware of where it returns Busy immediately, and the retries will show us if that's the case, but the busy_timeout won't. The retries will also show us if we just need to up the busy_timeout (though 30s should be more than enough, if we're doing everything else right).

As for the clones, I'm hoping this is the kind of thing rustc can optimize away, similar to how `Cow` is sort of irrelevant a lot of the time due to rustc optimizing away such clones. But again we'll remove that if no retries are actually needed.

Comment (freesig) [reply crates/holochain_sqlite/src/db.rs:300]: That's fair although you could achieve the same thing by setting the busy_timeout to 5 hours right? Because if you see any busy errors then you know that it's not from the busy timeout and it's some edge case that you are referring too? 
This is not your fault in anyway but it's a shame we have to change our code to suit a slow testing framework. Definitely something worth bringing up at the next stand up we are all at.

**Comment by maackle** [comment]
This didn't help
